### DISCUSSION TITLE
Cross Compiling on iMX.8X - SCALES

### DISCUSSION THREAD

**Discussion by kdizzlle** [nasa/fprime#3002] [Q&A]
@kevin-f-ortega @timcanham @Lex-ari 

## Problem Description

Trying to set up fprime on an 64-bit Linux host machine to cross compile c and c++ onto the phyCORE iMX.8X development board. Here is a link to our GitHub repo with our fprime environment: https://github.com/kdizzlle/fprime-hub-pattern-example

The iMX is using a poky Linux distribution with an SDK installed on the Linux host computer. The most relevant files for the SDK are in our repo as well.

We made our own cmake platform and toolchain, both called imx8x.cmake (in the repo). We tried to follow the pathing based on the I2C Interfacing directions on our docs: https://scales-hardware.readthedocs.io/en/latest/imx8x_procedures/#directions_3

Other relevant docs include the following:
phyCORE Install the SDK: https://docs.phytec.com/projects/yocto-phycore-imx8x/en/latest/applicationdevelopment/InstallTheSDK.html
phyCORE Hello World: https://docs.phytec.com/projects/yocto-phycore-imx8x/en/latest/applicationdevelopment/HelloWorld.html

We tried many different variations of the pathing for the platform and toolchain files, also inspired by the environment setup files from the SDK in the following repo: https://github.com/BroncoSpace-Lab/scales-hardware/tree/main/imx8x

Also in this repo is more screenshots that may be relevant. We really had to brute force it and disable cmake errors to get it to build even a little bit. 

## Context / Environment

Execute `fprime-util version-check` and share the output.

```
Operating System: Linux
CPU Architecture: x86_64
Platform: Linux-6.8.0-47-generic-x86_64-with-glibc2.35
Python version: 3.10.12
CMake version: 3.22.1
Pip version: 22.0.2
Pip packages:
    fprime-tools==3.4.4
    fprime-gds==3.4.3
    fprime-fpp-*==2.1.0a11
```

## Expected Behavior

Final build errors:

![fprime terminal output 7](https://github.com/user-attachments/assets/2a868c75-1ec5-4ff4-9734-35f5335db355)

**Comment by timcanham** [comment]
@kdizzlle I don't think we have permissions for your repo:
https://github.com/nasa/fprime/discussions/url

It looks like there are some platform-dependent implmentations in `SystemResources` that we need to fill in. In the short term, you could probably remove the `SystemResources` component from your build to make progress.

**Reply by kdizzlle** [reply]
Fixed the links in the post! Should work now.

**Comment by kevin-f-ortega** [comment]
Hi @kdizzlle, could you grant us read access to your repo?

**Reply by kdizzlle** [reply]
Fixed the links in the post! Should work now.

**Comment by kevin-f-ortega** [comment]
Hi Kelly. Thanks for fixing the links!  

Based on the warning, looks like you need to add `-O` to your compile options. Could you add that and try again?

Also, I noticed you added your platform and toolchain file inside of F Prime. I recommend moving them to your project's repo. So under https://github.com/kdizzlle/fprime-hub-pattern-example, create two directories called `cmake/platform` and `cmake/toolchain` and move the files into their respective folder. But we can tackle that later.

**Reply by kevin-f-ortega** [reply]
You'll add that compile option in your toolchain file

**Comment by kevin-f-ortega** [comment]
Also, as Tim mentioned, you could comment out SystemResources here:

1. https://github.com/kdizzlle/fprime-hub-pattern-example/blob/main/obcA/Top/instances.fpp#L135
2. https://github.com/kdizzlle/fprime-hub-pattern-example/blob/main/obcA/Top/topology.fpp#L43
3. https://github.com/kdizzlle/fprime-hub-pattern-example/blob/main/obcA/Top/topology.fpp#L112
4. https://github.com/kdizzlle/fprime-hub-pattern-example/blob/main/obcA/Top/obcAPackets.xml#L49-L79
5. and the same files for `obcB`

Once you comment out those lines, you can try to compile again.

**Reply by kevin-f-ortega** [reply]
Alternatively, could you try adding these lines into your platform file and compiling

```
choose_fprime_implementation(Os/Cpu Os/Cpu/Linux)
choose_fprime_implementation(Os/Memory Os/Memory/Linux)
```
